File: src/main/java/com/fendyk/clients/ClientAPI.java
package com.fendyk.clients;

import com.fendyk.API;

import java.util.HashMap;

public abstract class ClientAPI<FetchAPI, RedisAPI, Key, DTO> {
    protected static API api;
    protected FetchAPI fetch;
    protected RedisAPI redis;

    protected HashMap<Key, DTO> cachedRecords = new HashMap<>();

    public ClientAPI(FetchAPI fetch, RedisAPI redis) {
        this.fetch = fetch;
        this.redis = redis;
    }

    public FetchAPI getFetch() {
        return fetch;
    }

    public RedisAPI getRedis() {
        return redis;
    }


    /**
     * Returns the cached player, pure for UI/Visuals that require loads of updates only.
     * @param key
     * @return DTO
     */
    public DTO getCached(Key key) {
        return cachedRecords.get(key);
    }

    public static void setApi(API api) {
        ClientAPI.api = api;
    }
}

File: src/main/java/com/fendyk/clients/ClientAPI.kt
package com.fendyk.clients

abstract class ClientAPI<FetchAPI, RedisAPI, Key, DTO>(@JvmField var fetch: FetchAPI, @JvmField var redis: RedisAPI) {

    @JvmField
    protected var cachedRecords = HashMap<Key, DTO>()

    /**
     * Returns the cached player, pure for UI/Visuals that require loads of updates only.
     * @param key
     * @return DTO
     */
    fun getCached(key: Key): DTO? {
        return cachedRecords[key]
    }
}
